Ce/distribute ucr

snopoke opened the pull request.

calellowitz committed:
distribute delivery ucr

calellowitz committed:
Merge remote-tracking branch 'origin/sr-citus-co' into ce/distribute-ucr

calellowitz committed:
Merge remote-tracking branch 'origin/sk/temp-table-def' into ce/distribute-ucr

calellowitz committed:
fix bad merge

calellowitz committed:
fix pk migrations

calellowitz committed:
shorten table names

calellowitz committed:
add supervisor id to query

calellowitz committed:
fix test data

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
make ucrs distributed

calellowitz committed:
fix models for distribution

calellowitz committed:
begin fixing queries

calellowitz committed:
add supervisors to test data

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
distibute tables in tests and delete child tables

This commit will be removed eventually but make is easier to test things now

snopoke committed:
temp table helper

snopoke committed:
complementary feeding form aggregation

calellowitz committed:
fix complementary feeding ccs for citus

calellowitz committed:
fix child health thr for citus

calellowitz committed:
distribute gm ucr

snopoke committed:
remote temp table approach

snopoke committed:
make previous_month a sql param

snopoke committed:
fixups

snopoke committed:
wip agg gm forms

snopoke committed:
Merge branch 'sr-citus-co' into ce/distribute-ucr

snopoke committed:
filter prev_month by state_id

calellowitz committed:
distribute more UCRs

calellowitz committed:
update agg queries for citus

calellowitz committed:
update more agg models

calellowitz committed:
update agg script for citus

calellowitz committed:
add supervisor ids to test fixtures

calellowitz committed:
deduplicate form ids

calellowitz committed:
update monthly table queries for citus

calellowitz committed:
fix agg tables for citus

snopoke commented on custom/icds_reports/utils/aggregation_helpers/child_health_monthly.py:
```diff
@@ -54,14 +54,14 @@ def person_case_ucr_tablename(self):
 
     @property
     def tablename(self):
-        return "{}_{}".format(self.base_tablename, self.month.strftime("%Y-%m-%d"))
```
@calellowitz where was this table getting created before?

calellowitz replied:
In the tasks file. There was a sql file that did it.

calellowitz replied:
https://github.com/dimagi/commcare-hq/pull/23790/files#diff-b8c5a5e308a203bcaee8ed83e6837edfL471

snopoke committed:
make temp table unlogged

snopoke committed:
remove unused format params

snopoke committed:
remove unused function

snopoke commented on custom/icds_reports/utils/aggregation_helpers/child_health_monthly.py:
```diff
@@ -335,7 +343,10 @@ def pre_aggregation_queries(self):
     def create_temporary_table(self):
-        return "CREATE TABLE \"{}\" (LIKE child_health_monthly INCLUDING INDEXES)".format(self.temporary_tablename)
+        return """
+        CREATE UNLOGGED TABLE \"{table}\" (LIKE child_health_monthly INCLUDING INDEXES);
```
@calellowitz FYI any temporary tables that we create that aren't real temp tables should be 'unlogged' which give much better performance.

snopoke commented on custom/icds_reports/utils/aggregation_helpers/agg_child_health.py:
```diff
@@ -134,46 +134,55 @@ def aggregation_query(self):
     def update_queries(self):
         yield """
-            UPDATE "{tablename}" agg SET
+        CREATE TEMPORARY TABLE "{tmp_tablename}" AS SELECT
```
is this necessary here? neither the table being updated (`agg_child_health....`) and the table being selected from (`awc_location`) are local tables so I don't think the temp table approach is necessary here

snopoke replied:
offlined - `awc_location` is a reference table - some uncertainty about that but ok for now

snopoke committed:
remove unused formatting var

snopoke committed:
rename format vars

snopoke committed:
merge master

snopoke committed:
Merge branch 'sr-citus-co' into ce/distribute-ucr

snopoke committed:
fix migrations

snopoke committed:
fix migration

snopoke committed:
distribute tables

snopoke committed:
distribute static-awc_location

calellowitz committed:
modify agg awc queries for citus

calellowitz committed:
fix ccs and ch mbt

snopoke committed:
distribute ccs_record_cases_monthly

snopoke committed:
remove duplicate method

snopoke committed:
distribute daily_feeding_forms

snopoke committed:
remove rows with null child_health_case_id and null supervisor ID

Not 100% sure if this is the correct way to fix these rows
with missing supervisor IDs

snopoke committed:
distribute awc_mgt_forms

snopoke committed:
month is already at day 1

snopoke committed:
mark supervisor ID column as deprecated

snopoke committed:
distribute ls_vhnd and ls_beneficiary

snopoke committed:
distribute child_delivery_forms

snopoke committed:
distribute daily_attendance

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz commented on custom/icds_reports/ucr/data_sources/ccs_record_cases_monthly_tableau2.json:
```diff
@@ -3136,13 +3136,11 @@
-          "constraint": "month"
-        }
-      ]
+      "citus_config": {
```
This table is no longer used in the aggregation and will likely be deleted before we roll out citus. Not sure if there is a downside to having it distributed for now, but also not sure it is necessary

calellowitz committed:
fix mbt queries for citus

calellowitz committed:
remove unnecessary join condition

calellowitz committed:
change tests to work with new tables

calellowitz committed:
remove print

calellowitz committed:
fix issues in migrations

calellowitz committed:
fix state condition in multi month query

calellowitz committed:
fix inconsistency in person ucr fixture

calellowitz committed:
fix inactive aww queries for citus

calellowitz committed:
proper child health filtering in tests

snopoke committed:
move code into helpers

snopoke committed:
add distributed helpers from master

snopoke committed:
copy new code to distributed helpers

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
Merge branch 'sk/icds-citus' into ce/distribute-ucr

snopoke committed:
rename

snopoke committed:
remove unchanged helpers

snopoke committed:
fix migration conflict

snopoke committed:
add distributed helpers

snopoke committed:
better system for getting helper

calellowitz committed:
fix test regressions

calellowitz committed:
add new distributed helpers

calellowitz committed:
add local awc_location

snopoke committed:
fix circular imports with mbt

snopoke committed:
remove unused functions

snopoke committed:
simpler helper setup

snopoke committed:
add new helpers

snopoke committed:
sort

snopoke committed:
remove dupicate field

snopoke committed:
consolidate base classes

snopoke committed:
call function

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
add local location tables and views

calellowitz committed:
use local location view in ccs view

calellowitz committed:
fix incentive report for citus

snopoke committed:
remove get_view_migrations from squashed

snopoke committed:
create location view before it's used

snopoke committed:
Merge branch 'sk/add-composite-key-migration' into ce/distribute-ucr

calellowitz committed:
update rest of views to use local table

calellowitz committed:
create table if not exists

calellowitz committed:
use local location table for mbt

calellowitz committed:
add ordering for performance report download

calellowitz committed:
change left join to inner in dashboard views

calellowitz committed:
fix new born lbw report

calellowitz committed:
undistribute ccs record monthly ucr

calellowitz committed:
python2/3 compat for module exports

snopoke committed:
[ci skip] stickler

snopoke committed:
small fixups

snopoke committed:
create new tables for monolith DB

snopoke commented on custom/icds_reports/ucr/data_sources/awc_locations.json:
```diff
@@ -67,7 +67,7 @@
         "transform": {},
         "is_nullable": false,
-        "is_primary_key": false,
+        "is_primary_key": true,
```
don't think this is still needed

snopoke reviewed (commented):
Did a review and made some notes regarding the merge path for this:

* Create and populate `icds_months_local` and `awc_location_local` prior to merge
* QA view join changes (`ccs_record_monthly`, `child_health_monthly`)
* Missing migration for CcsRecordMonthly primary key change
* Missing migration for ChildHealthMonthly primary key change
* Create daily_attendance index on awc_id prior to merge

Also some changes from master that need merging.

calellowitz committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
remove extra primary key

calellowitz committed:
fix migrations

calellowitz committed:
create pk for monthly tables

calellowitz commented:
> Create and populate icds_months_local and awc_location_local prior to merge

There is already a migration to create these tables, and I set it up so they get populated during the aggregation, so I think we may be good to go on these.

> Missing migration for CcsRecordMonthly primary key change
> Missing migration for ChildHealthMonthly primary key change

Added these and fixed the migration ordering. I think that is all the code changes you mentioned. The rest is just rollout.

snopoke commented:
>> Create and populate icds_months_local and awc_location_local prior to merge

> There is already a migration to create these tables, and I set it up so they get populated during the aggregation, so I think we may be good to go on these.

@calellowitz my concern with this is that the tables won't be populated until after the first aggregation but they will be used immediately which will make it look like data is missing

calellowitz commented:
Fair enough. I expected this would get taken care of if we were running in parallel at all

snopoke committed:
use util functions

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
integrate recently merged changes

snopoke committed:
distribute tables

snopoke committed:
fix migrations

snopoke committed:
run tests without citus

snopoke committed:
add test for agg helper output

snopoke committed:
trim whitspace

snopoke committed:
further migration fix

snopoke committed:
stickler

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
update distributed queries and diffs

snopoke committed:
add encoding to mbt

snopoke committed:
cleanup

snopoke committed:
use py2 func

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
order params

snopoke committed:
fix migrations

snopoke committed:
drop if exists

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
allow parallel aggregations

calellowitz committed:
fix typo

calellowitz committed:
lint

sravfeyn committed:
Use Citus for ICDS dashboard view when user has toggle

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
incorporate changes from master

https://github.com/dimagi/commcare-hq/pull/24044

snopoke committed:
update diffs

snopoke committed:
fix migrations

calellowitz committed:
force citus for helpers

calellowitz committed:
add missing supervisor column

calellowitz committed:
fix table name typo

calellowitz committed:
do not distribute incentive or infra tables

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
incorporate changes from master

https://github.com/dimagi/commcare-hq/pull/24089

snopoke committed:
fix create table

snopoke committed:
don't use citus for tests yet

calellowitz committed:
fix syntax errors in pnc query

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
update diffs

snopoke committed:
fix migrations

calellowitz committed:
force citus in subtasks

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
fix migrations

snopoke committed:
fix db name

snopoke committed:
update diffs

snopoke committed:
use delete instead of truncate

snopoke committed:
fix diffs

snopoke committed:
fix var name reuse lint

snopoke committed:
better db name

snopoke committed:
fix ordering in test

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
update migration

emord commented on custom/icds_reports/ucr/data_sources/child_delivery_forms.json:
```diff
@@ -83,8 +83,8 @@
         "transform": {},
-        "is_nullable": true,
-        "is_primary_key": false,
+        "is_nullable": false,
```
Yeah. that would. I'm not sure how many delivery forms there are, but I imagine building a new data source would be fairly quick. i.e. create delivery_forms_v2 with is_nullable: true, rebuild that, then move all references to child_delivery_forms to the v2 verson

calellowitz commented on custom/icds_reports/ucr/data_sources/ccs_record_cases.json:
```diff
@@ -125,7 +125,7 @@
         "type": "expression",
         "transform": {},
         "is_nullable": true,
```
yeah, it should. `person_cases.csv` has the same issue. based on the discussion on the delivery forms data source, it sounds like this will be difficult to change?

calellowitz committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
add missing supervisor ids

used following awk command then replaced the file
`awk -F, -v OFS=, 'NR == FNR {s[$1] = $5} NR > FNR {if ($4 == "") {$4 = s[$3]} print}' location_ucr.csv person_cases.csv > person_cases.csv.new`

snopoke committed:
set nullable to false where primary key is true

snopoke commented:
@calellowitz @emord regarding the nullable column change, I was able to check all but 3 tables and apply the column change directly to the table which will prevent a rebuild. The tables I was not able to check are:

file name (.json) | table name
-- | --
dashboard_daily_feeding_forms | ucr_icds-cas_dashboard_child_health_daily_2cd9a7c1
dashboard_growth_monitoring | ucr_icds-cas_static-dashboard_growth_moni_4ebf0625
postnatal_care_forms | ucr_icds-cas_static-postnatal_care_forms_0c30d94e

These all have a lot of child tables and a lot of data so querying them directly was taking too long. I think we could try querying each child table individually to find rows with null values. It will also take a long time to apply the `not null` change. I'm not sure if there is a way around this. Perhaps we can leave it off for now and only apply it when we migrate the data to CitusDB?

emord commented:
>  It will also take a long time to apply the not null change

@snopoke what about changing the migration code to something like, if we need to add a constraint and the table already exist, we don't, but on table creation we add the constraint? I'm not sure how difficult that is.

Might be worth creating a view where you can find discrepanices between tables and definitions though.

if we want to catch nulls on already existing data sources we could add it to the data source validations

calellowitz committed:
integrate delivered in month changes

calellowitz committed:
update ccs record monthly diffs

calellowitz committed:
no null supervisors

calellowitz committed:
update ccs record monthly diffs

calellowitz committed:
fix wrong property

calellowitz committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
update ccs record monthly diffs

calellowitz commented:
✅ on the build!

calellowitz committed:
Merge branch 'ce/distribute-ucr' into ce/parallel-agg-local

calellowitz committed:
Merge pull request #24113 from dimagi/ce/parallel-agg-local

Ce/parallel agg local

calellowitz committed:
Merge branch 'ce/distribute-ucr' into sr-citus-toggle

calellowitz committed:
use citus override in middleware

snopoke committed:
run composite key migrations on citusdb

snopoke committed:
add missing migration for AwcLocationLocal

snopoke committed:
distribute tables in migration

snopoke committed:
always add composite keys in tests

snopoke committed:
distribute UCRS

snopoke committed:
Merge pull request #24117 from dimagi/sr-citus-toggle

Use Citus for ICDS dashboard view when user has toggle

snopoke committed:
[ci skip] import

calellowitz committed:
fix location inconsistency in test data

calellowitz committed:
Merge branch 'master' into ce/distribute-ucr

snopoke committed:
fix migrations

snopoke committed:
include ID in error

snopoke committed:
support django curosr

snopoke committed:
format line

snopoke committed:
swap order for correct diff output

snopoke committed:
order of filters does not matter

snopoke committed:
fix ucr configs

snopoke committed:
fix queries on household table

snopoke committed:
fix usage forms queries

snopoke committed:
update sql output

snopoke committed:
add supervisor_id to usage.csv

snopoke committed:
Merge branch 'master' into ce/distribute-ucr

calellowitz committed:
few more test fixes

kaapstorm commented:
:boom:
